### ISSUE TITLE
Further merge --script-* with --*-hook

### ISSUE THREAD

**Issue by pde** [certbot/certbot#3767]
Today @jsha made an argument that we should work a bit harder to make the script plugin feel and operate consistently with the existing hooks.  We reached some conclusions about this:

* `--script-auth` and `--script-cleanup` could become `--auth-hook` and `--cleanup-hook`; that would require moving them out of a plugin but that's doable.
* Support from them could be moved from their own plugin to the manual plugin, with the advantage that existing manual users could told about them interactively and onboarded into the brave new scripting future
* We disagreed on whether `--auth-hook` should always be run or whether it should be specific to `--manual`, but concluded that this would be the prerogative of whoever implements this refactor.

**Comment by bmw** [comment]
I was going to take this unless someone else wanted to.

**Comment by joohoi** [comment]
It's a hard question. There certainly are benefits to the both approaches, however I would suggest leaving it as is - to script plugin. This is not up to me, but here are my thoughts about the subject to help you to pick the way to proceed.

Things that would promote merging the functionality with hooks:
 - Code simplicity
 - Being able to possibly use auth and cleanup hooks from inside other plugins also if needed

However things that I see beneficial for having the functionality inside a plugin, be it merged with manual plugin, or left to script plugin:
 - Keeping the authenticator / installer functionality inside plugin, and staying consistent with current design. I think this is also intuitive from the user perspective, as the current hooks are pretty much "one shot" scripts, that are run outside of the actual auth loop, ie: outside of the "magic".
 - In contrary to hooks, plugins are for the actual "heavy lifting" process, while hooks are more like helpers in my view. This is effectively what the authenticator part is doing.

Things promoting keeping it in script plugin instead of merging with manual:
 - The current manual plugin users have either scripted around it or have some kind of process around it. I don't think there are actual conversion factors.
 - Omitting the current manual plugin pause-to-wait-for-challenge-deployment functionality would break scripts for users that have scripted around the manual plugin altogether.

**Comment by bmw** [comment]
I wanted to comment on some of the things that have been said here.

> We disagreed on whether --auth-hook should always be run or whether it should be specific to --manual, but concluded that this would be the prerogative of whoever implements this refactor.

I have a strong preference for making the hooks specific to the manual plugin. Doing otherwise will result in additional complexity both in the implementation and in the user experience. It is additional complexity for the implementation because the Certbot core needs to call into multiple places for both authentication and cleanup and it complicates error handling. It is additional complexity for the user as I don't think there is an intuitive answer to whether the hooks should run before or after the plugin they run with and it creates the strange situation where people are forced to choose an arbitrary plugin that does work they don't care about only to use the auth and cleanup hooks. This view is not inconsistent with what Peter wrote here, assuming I'm the one doing the work and I'll implement this as desired. I brought this up because it informs my next point.

> --script-auth and --script-cleanup could become --auth-hook and --cleanup-hook; that would require moving them out of a plugin but that's doable.

Because I think these flags should only be used for one plugin, I think it is a very bad idea to not differentiate them from the other hooks in some way. For this reason, I was planning on implementing them as `--manual-auth-hook` and `--manual-cleanup-hook`. Perhaps I should have been more vocal about my disagreement here as I definitely do not agree we should just call them `--auth-hook` and `--cleanup-hook`.

My options in order of preference are to name these flags:

1. `--<plugin>-<verb>-hook`
2. Not call them hooks at all
3. `--auth-hook` and `--cleanup-hook`

These flags are essentially hooks provided by the user, however, because I think they should only be applied to one plugin, they should be clearly differentiated in some way. I like the first option, because I think it clearly says they are hooks for that plugin. If we don't add a plugin name prefix, I do not agree that these options should be called hooks.

> The current manual plugin users have either scripted around it or have some kind of process around it. I don't think there are actual conversion factors.

Conversation factors may be minimal. The main reason this was pushed for was to reduce to the number of plugins that Certbot ships with. @jsha's comment that I agree with was that plugin choice is probably the hardest decision that Certbot users have to make. Let's not add more options unless it is necessary.

> Omitting the current manual plugin pause-to-wait-for-challenge-deployment functionality would break scripts for users that have scripted around the manual plugin altogether.

I planned on implementing this effectively like this for most methods in the plugin:
```
def foo():
    if self.conf('auth') and self.conf('cleanup'):
        script_plugin_foo()
    else:
        manual_plugin_foo()
```
The actual implementation is likely to be cleaner/cleverer, but that is how I see the behavior. Perhaps both hooks shouldn't be required, but my basic idea was to have two completely separate code paths. The pause to wait functionality would not be changed.

> I think the functionality is very beneficial for larger integrations, so I think it should be promoted on plugin level instead of "hiding" in the concept of hooks.

I agree this functionality is a big win for a lot of people given the current complexity of writing a Certbot plugin. I think these hooks are only "hidden" for existing Certbot users. Like we're experiencing for the Nginx plugin, many people don't realize that the plugin has changed. For new users, however, and for the long term, big picture of Certbot, I don't think there is a clear win to having this functionality in a separate plugin. I think we can be very clear in Certbot's help, documentation, and UI that the manual plugin is the DIY plugin, whether that be writing your own scripts or performing the task by hand.

> It would be counterintuitive to run manual (or any plugin in that effect) as a null plugin, functioning only as a container for auth-hook.

I completely agree that running a plugin as a null plugin just to get the hooks is all kinds of awkward as I outlined in my first section in this post. I'm not sure if this changes your opinion here, but as I said above I basically want to change `manual`'s behavior when an auth hook is provided. The normal functionality of `manual`, which is of minimal value with the addition of auth hooks, would not be executed when an auth hook is given on the command line.

If anyone would like to express any additional thoughts on this before I start implementing this, I would love to hear them.

**Comment by joohoi** [comment]
Thanks for your essay on the subject @bmw ;) That gave me a better view on problems with adding plugins and your plans about the merging. I can agree with your points there, and I see no problem with merging functionality to manual plugin if the actual functions are not hidden behind hooks.

**Comment by bmw** [comment]
Fixed in #3890.
